### PR TITLE
Add support for php 8.1

### PR DESCRIPTION
**Pull request by marko-ilic**
In the scope of this PR, the support for PHP 8.1 is added.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,6 +5,7 @@ php:
   - 7.3
   - 7.4
   - 8.0
+  - 8.1.0
```
Comment (reedy) [inline .travis.yml:8]: Why is this 8.1.0? Why not just 8.1? 8.1.4 is out already...

File: README.md [hunk 2]
```diff
@@ -19,7 +19,7 @@ A full history of the different versions of Diff can be found in the [release no
 
 **Diff 3.x:**
 
-* PHP 7.0 or later (tested with PHP 7.0 up to PHP 7.4)
+* PHP 7.2 or later (tested with PHP 7.2 up to PHP 8.1)
```
Comment (reedy) [inline README.md:22]: Yay, out of date docs

File: src/DiffOp/Diff/Diff.php [hunk 3]
```diff
@@ -145,21 +145,23 @@ private function preSetElement( $index, DiffOp $value ): bool {
 	 *
 	 * @param string $serialization
 	 */
-	public function unserialize( $serialization ) {
-		$serializationData = unserialize( $serialization );
+	public function unserialize( $serialization ): void {
+		$this->__unserialize( unserialize ($serialization) );
```
Comment (reedy) [inline src/DiffOp/Diff/Diff.php:149]: extra space after function call. needs spaces inside `()`

File: src/DiffOp/Diff/Diff.php [hunk 4]
```diff
@@ -145,21 +145,23 @@ private function preSetElement( $index, DiffOp $value ): bool {
...
+	}
 
-		foreach ( $serializationData['data'] as $offset => $value ) {
+	public function __unserialize( array $data ): void {
```
Comment (reedy) [inline src/DiffOp/Diff/Diff.php:152]: Probably want to copy docs across

File: src/DiffOp/DiffOpAdd.php [hunk 5]
```diff
@@ -64,8 +69,12 @@ public function serialize() {
 	 *
 	 * @param string $serialization
 	 */
-	public function unserialize( $serialization ) {
-		$this->newValue = unserialize( $serialization );
+	public function unserialize( $serialization ): void {
+		$this->__unserialize( unserialize ($serialization) );
+	}
+
+	public function __unserialize( array $serialization ): void {
```
Comment (reedy) [inline src/DiffOp/DiffOpAdd.php:76]: Copy docs here too

File: src/DiffOp/DiffOpChange.php [hunk 6]
```diff
@@ -65,8 +65,13 @@ public function getNewValue() {
 	 *
 	 * @return string|null
 	 */
+	#[\ReturnTypeWillChange]
 	public function serialize() {
-		return serialize( [ $this->newValue, $this->oldValue ] );
+		return serialize( $this->__serialize() );
+	}
+
+	public function __serialize(): array {
```
Comment (reedy) [inline src/DiffOp/DiffOpChange.php:73]: And here

File: src/DiffOp/DiffOpChange.php [hunk 7]
```diff
@@ -76,8 +81,12 @@ public function serialize() {
 	 *
 	 * @param string $serialization
 	 */
-	public function unserialize( $serialization ) {
-		list( $this->newValue, $this->oldValue ) = unserialize( $serialization );
+	public function unserialize( $serialization ): void {
+		$this->__unserialize( unserialize ($serialization) );
+	}
+
+	public function __unserialize( array $serialization ): void {
```
Comment (reedy) [inline src/DiffOp/DiffOpChange.php:88]: And here

File: src/DiffOp/DiffOpRemove.php [hunk 8]
```diff
@@ -64,8 +69,12 @@ public function serialize() {
 	 *
 	 * @param string $serialization
 	 */
-	public function unserialize( $serialization ) {
-		$this->oldValue = unserialize( $serialization );
+	public function unserialize( $serialization ): void {
+		$this->__unserialize( unserialize ($serialization) );
+	}
+
+	public function __unserialize( array $serialization ): void {
```
Comment (reedy) [inline src/DiffOp/DiffOpRemove.php:76]: And here

File: src/DiffOp/Diff/Diff.php [hunk 9]
```diff
@@ -383,7 +385,7 @@ private function getNewOffset(): int {
 	 *
 	 * @param mixed $value
 	 */
-	public function append( $value ) {
+	public function append( $value ): void {
```
Comment (reedy) [inline src/DiffOp/Diff/Diff.php:388]: Is this related?

File: src/DiffOp/Diff/Diff.php [hunk 10]
```diff
@@ -395,7 +397,7 @@ public function append( $value ) {
 	 * @param int|string $index
 	 * @param mixed $value
 	 */
-	public function offsetSet( $index, $value ) {
+	public function offsetSet( $index, $value ): void {
```
Comment (reedy) [inline src/DiffOp/Diff/Diff.php:400]: Is this related?

**Review by reedy — CHANGES_REQUESTED** [review]
Thanks! :)

File: .travis.yml (hunk 1 above)
Comment (marko-ilic) [reply .travis.yml:8]: This is a Travis issue. Afaik only works with 8.1.0. 
For example [8.1.3](https://app.travis-ci.com/github/wmde/Diff/jobs/563855865)

File: src/DiffOp/Diff/Diff.php (hunk 9 above)
Comment (marko-ilic) [reply src/DiffOp/Diff/Diff.php:388]: Yes, in PHP 8.1 deprecation warning will be thrown 
`Deprecated: Return type of Diff\DiffOp\Diff\Diff::append($value) should either be compatible with ArrayObject::append(mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /Users/milic/Development/Diff/src/DiffOp/Diff/Diff.php on line 395`

File: src/DiffOp/Diff/Diff.php (hunk 10 above)
Comment (marko-ilic) [reply src/DiffOp/Diff/Diff.php:400]: The [same](https://github.com/wmde/Diff/pull/129#discussion_r843827883) like for `append()`

File: .travis.yml (hunk 1 above)
Comment (reedy) [reply .travis.yml:8]: Aha, ok then. Annoying, but hopefully something they fix later...

File: src/DiffOp/Diff/Diff.php (hunk 9 above)
Comment (reedy) [reply src/DiffOp/Diff/Diff.php]: Wouldn't using `#[\ReturnTypeWillChange]` make more sense for the time being then? At least, until we can drop support for older PHP versions?

Comment (marko-ilic) [reply src/DiffOp/Diff/Diff.php]: 👍

File: src/DiffOp/Diff/Diff.php (hunk 10 above)
Comment (reedy) [reply src/DiffOp/Diff/Diff.php:409]: `#[\ReturnTypeWillChange]` here too?

**Comment by reedy** [comment]
Some minor code style issues too

```
FILE: /home/travis/build/wmde/Diff/src/DiffOp/Diff/Diff.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 396 | ERROR | [x] Tabs must be used to indent lines; spaces are not
     |       |     allowed
     |       |     (Generic.WhiteSpace.DisallowSpaceIndent.SpacesUsed)
----------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```

**Comment by reedy** [comment]
I think this is nearly ready to merge :)

File: .travis.yml (hunk 1 above)
Comment (reedy) [reply .travis.yml:8]: https://travis-ci.community/t/php-8-1-support/12439/23 for future reference

https://docs.travis-ci.com/user/languages/php/#php-versions seems to suggest there is a 8.1 and an 8.1.4 (but no 8.1.3)..

**Comment by reedy** [comment]
LGTM, probably ready to squash and merge. Which I unfortunately do not have rights on this repo to do.

I've just created https://github.com/wmde/Diff/pull/130 to play around with testing PHP 8.1 etc in `.travis.yml`

File: .travis.yml (hunk 1 above)
Comment (reedy) [reply .travis.yml:8]: Correction. There's `8.1` on 16.04... Not so helpful.

There seems to be a mismatch of random PHP versions on different Ubuntu releases. 20.04 and 18.04 have `8.1.0` and `8.1.2`. 16.04 has `8.1`, `8.1.0`, `8.1.2`, `8.1.4`.

Might be "slightly" better to use `8.1.2` as the latest point release available.

File: src/DiffOp/Diff/Diff.php (hunk 4 above)
Comment (lucaswerkmeister) [reply src/DiffOp/Diff/Diff.php:160]: I don’t think copying `@see Serializable::unserialize` for an `__unserialize` method makes much sense, it’s not the same method…

File: src/DiffOp/DiffOpAdd.php [hunk 11]
```diff
@@ -64,8 +76,20 @@ public function serialize() {
 	 *
 	 * @param string $serialization
 	 */
+	#[\ReturnTypeWillChange]
 	public function unserialize( $serialization ) {
-		$this->newValue = unserialize( $serialization );
+		$this->__unserialize( unserialize ($serialization) );
```
Comment (lucaswerkmeister) [inline src/DiffOp/DiffOpAdd.php:81]: Isn’t this an incompatible change to the serialization? Previously, `unserialize( $serialization )` was directly the `newValue`, which could apparently be anything; now, it’s expected to be an array *containing* the `newValue`. If the new code tries to deserialize a value serialized by the old code, what happens?

**Review by lucaswerkmeister — COMMENTED** [review]
Provocative question: do we actually *need* this class to be serializable at all? Or can we make our lives easier by just dropping the serialization code? (This would be a breaking change, to be made in a 4.0 release, of course.)

I did a test in my local wiki, and as far as I can tell, everything still worked with the serialization code ripped out (manually edited in `vendor/` – yes, I made sure there weren’t any other copies of the library in nested vendor directories), including change dispatching to a second wiki. I also can’t think of a situation where we’d need to serialize diff objects (whereas data model objects *are* serialized, as we’ve established in [T301249](https://phabricator.wikimedia.org/T301249) / [I7f5c0f125d](https://gerrit.wikimedia.org/r/c/mediawiki/extensions/Wikibase/+/773521)).

File: src/DiffOp/DiffOpAdd.php (hunk 11 above)
Comment (marko-ilic) [reply src/DiffOp/DiffOpAdd.php:79]: Isn't the same case here? https://github.com/wikimedia/mediawiki/commit/4cc1a3eecaaecc4368753a9a08fa0732bddae6fb

Comment (lucaswerkmeister) [reply src/DiffOp/DiffOpAdd.php:79]: Do you have a specific class in mind? At a glance, I don’t see the same problem in that commit.

**Comment by jdforrester** [comment]
Hello lovely people. This is now critical-path for running MediaWiki tests on PHP 8.1 (due to Wikibase's privileged position in the gate). Is there any chance this could be completed and released so we can move forward?

**Comment by mariushoch** [comment]
This still needs to be resolved: https://github.com/wmde/Diff/pull/129/files#r846253808 (and the same in [DiffOpRemove](https://github.com/wmde/Diff/pull/129/files#diff-793d3e2df946c91998a21512599267a811e64b2ef3a440b1a042ed689375e91f)).

Other than that, I think this is good to go.
